### PR TITLE
Add Transaction Building guide

### PR DESCRIPTION
**Pull request by capt-nemo429**

### CODE REVIEW

File: docs/transaction-building.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+```
+
+By default, the `build()` method will return a default unsigned transaction object, if you want to send it to be signed by the user's wallet through [dApp Connector protocol](./wallet-interaction.md), you should include "EIP-12" as a parameter in the `build()` method so it will include additional information required by the protocol.
+
+<!-- prettier-ignore -->
+```ts
+new TransactionBuilder(creationHeight)
+   // ...
+  .build("EIP-12"); // [!code focus]
+```
```
Comment (cafebedouin) [inline docs/transaction-building.md:220]: Probably should either have a link to the total built transaction or have it on the end as an appendix, so someone can copy and paste the whole transaction rather than need to piece it together. This tutorial assumes someone is following a step-by-step process. If this were me, I'd want to copy the whole, read this to understand it, make modifications, then run it to see if I am using it right using a test wallet. I imagine that is probably the norm.

**Review by cafebedouin — COMMENTED** [review]
This is a first run through. I normally would like to do a few passes, but I'm going to go with the view that it is better to do more reviews rather than trying to up the quality of any individual review. Also, I tend to have a heavy hand with editing, and may have gotten some of the programming details wrong. Let me know if I need to do something different.

Also, thanks for sending the screen recording. I haven't done this before, and it was helpful.

File: docs/transaction-building.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+Each newly created output box needs to hold its creation block height. Usually, the Creation Height is the current block height at the time of the creation of the transaction.
+
+```ts
+new TransactionBuilder(creationHeight);
+```
+
+## Step. 2: Add Inputs
+
+Input boxes are the source of the funds for the transaction. As they hold the funds to be spent by the transaction, the guard script, also known as `ErgoTree` in Ergo, in each input box must be successfully evaluated by the nodes for the transaction to be considered valid.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:39]: Yeah, you have a good point here.

File: docs/transaction-building.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+- Each transaction must contain **one or more** Input boxes, **one or more** Output boxes and **zero or more** Data-Input boxes.
+
+:::warning
+**Work-in-Progress**: Non-reviewed text. You may find numerous writing errors throughout this guide.
+:::
+
+## What is a box?
+
+A Box is an **Unspent Transaction Outputs (UTxO)** on steroids it's not simply a coin as it's commonly intended in blockchains like Bitcoin. Beyond protecting monetary values with a complex script (smart contract), it can also contain data, tokens, and registers. That's what eUTxO stands for.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:15]: I think we have some misconceptions here, lemme try to clarify it a bit using your suggestions

File: docs/transaction-building.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+An Ergo Transaction is nothing more than a set of boxes.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:17]: The intention behind this statement is exactly to wrap in a few words what is a transaction, so I think we might should keep it as it is.

File: docs/transaction-building.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+## Building a transaction
+
+For transaction building, Fleet provides the `TransactionBuilder` and the `OutputBuilder`. Both deliver a seamless way to construct transactions with built-in validations, selection strategies, automatic change calculation, _et cetera_.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:21]: Nice catch! It's a class, actually. Will fix it.

File: docs/transaction-building.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+```ts
+import { TransactionBuilder, OutputBuilder } from "@fleet-sdk/core";
+```
+
+## Step. 1: Instantiating and adding the Creation Height
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:27]: We cannot instantiate a transaction, but the TransactionBuilder, so for the sake if conciseness I choose to use it just as "Instantiating", which is sub-intended to be used with a class.

File: docs/transaction-building.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+<!-- prettier-ignore -->
+```ts
+new TransactionBuilder(creationHeight)
+  .from(inputs);
+```
+
+### Input box object type
+
+You can fetch output boxes from the blockchain or directly from the [user's wallet](./wallet-interaction.md#step-5-fetch-boxes).
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:51]: Will rework this.

File: docs/transaction-building.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+  creationHeight: number;
+  additionalRegisters: NonMandatoryRegisters;
+  index: number;
+  transactionId: TransactionId;
+};
+```
+
+## Step. 3: Add Data-Inputs
+
+The Data-Inputs are boxes whose data can be referenced and used by smart contracts of the input boxes if required. A good use case for Data-Inputs is oracle data feeding.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:68]: Great addition!

File: docs/transaction-building.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+## Step. 4: Add Outputs
+
+Outputs boxes are the destination of funds held by the input boxes. When confirmed by the blockchain, output boxes are ready to be used as input boxes in further transactions.
+
+In the following example, we are using the `OutputBuilder` and the ` TransactionBuilder`'s `to()` method to construct and include outputs in the transaction body.
+
+The `to()` method requires one or more `OutputBuilder` objects to be included in the transaction body.
+
+The `OutputBuilder` provides a handy set of methods for output construction, such as token inclusion, register set, token minting, and so on. To be instantiated, the `OutputBuilder` requires two construction parameters: the amount of `nanoergs` that should be put in the output box and the recipient address.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:91]: Yeah, that will be automatically generated from code.

File: docs/transaction-building.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+Each transaction on Ergo needs to have a mining fee box, you can use the `payFee()` method to specify the fee amount in `nanoergs`.
+
+<!-- prettier-ignore -->
+```ts
+new TransactionBuilder(creationHeight)
+  // ...
+  .payFee("1100000"); // [!code focus]
+```
+
+Alternatively, you can use the `payMinFee()` method to add the min recommended miner fee amount, which is `1100000 nanoergs` or `0.0011 ERG`.
```
[the remark this answers is not in the document]
Comment (capt-nemo429) [reply docs/transaction-building.md:191]: It's a nice suggestion, will add.

File: docs/transaction-building.md (hunk 1 above)
Comment (capt-nemo429) [reply docs/transaction-building.md:220]: Makes very sense, will add.

File: docs/transaction-building.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+- Each transaction must contain **one or more** Input boxes, **one or more** Output boxes and **zero or more** Data-Input boxes.
+
+:::warning
+**Work-in-Progress**: Non-reviewed text. You may find numerous writing errors throughout this guide.
+:::
+
+## What is a box?
+
+Shortly a box is an **Extended Unspent Transaction Output (eUTxO)**, which extends the Bitcoin's **Unspent Transaction Output (UTxO)** data structure with tokens, registers, data, and smart contracts.
```
Comment (cafebedouin) [inline docs/transaction-building.md:15]:
```suggestion
Briefly stated, a box is an **Extended Unspent Transaction Output (eUTxO)** data structure that extends Bitcoin's **Unspent Transaction Output (UTxO)** beyond coins to include tokens, registers, data, and smart contracts.
```

File: docs/transaction-building.md [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+An Ergo Transaction is nothing more than a set of boxes.
```
Comment (cafebedouin) [inline docs/transaction-building.md:17]:
```suggestion
An Ergo Transaction is nothing more than changing boxes, inputs into outputs.
```

Comment (cafebedouin) [reply docs/transaction-building.md:17]: The transaction is making a change. I feel this is important to include. An address can have a set of boxes, but it is not a transaction. The defining characteristic of a transaction is it changes the blockchain, which means changing boxes. Maybe pedantic, but it seems clearer to me. Maybe there's another way to say it too?

File: docs/transaction-building.md [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+};
+```
+
+## Step. 3: Add Data-Inputs
+
+The Data-Inputs are boxes whose data can be referenced and used by contracts, if required. A good use case for Data-Inputs is using oracle data. For example, the [SigmaUSD](https://sigmausd.io/) contract uses an oracle of Ergo price to set the exchange rate for SigUSD and SigRSV conversation to and from ERG at regular intervals.
+
+- Any **unspent** box can be used as Data-Input.
+- Data-Inputs are **optional** and must be only included in the transaction if required by a contract.
+- Boxes will not be spent by the transaction if used as Data-Input.
```
Comment (cafebedouin) [inline docs/transaction-building.md:76]:
```suggestion
- Boxes will not be spent by the transaction if used as Data-Input in a transaction.
```

File: docs/transaction-building.md [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+new TransactionBuilder(creationHeight)
+  .from(inputs)
+  .withDataFrom(dataInputs);
+```
+
+## Step. 4: Add outputs
+
+Outputs boxes are the destination of funds held by the input boxes. When confirmed by the blockchain, output boxes are ready to be used as input boxes in further transactions.
+
+In the following example, we are using the `OutputBuilder` and the ` TransactionBuilder`'s `to()` method to construct and include outputs in the transaction body.
```
Comment (cafebedouin) [inline docs/transaction-building.md:91]:
```suggestion
In the following example, the `OutputBuilder` and the ` TransactionBuilder`'s `to()` methods are used to construct and include outputs in the transaction body.
```

Comment (cafebedouin) [reply docs/transaction-building.md:91]: Normally, people suggest avoiding the passive voice, but I think I prefer it here because of the step-by-step of the tutorial. You don't want to put yourself between the user and the actions the document is suggesting.

File: docs/transaction-building.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        tokenId: "1fd6e032e8476c4aa54c18c1a308dce83940e8f4a28f576440513ed7326ad489", // [!code focus]
+        amount: "100" // [!code focus]
+      }) // [!code focus]
+  );
+```
+
+:::tip
+If you are trying to add a token that may already be present in the output's token list, the default behavior of the `addTokens()` method is to sum the amount of duplicated tokens and keep a single record per token.
+
+If you want it to be added at the end of the list, you can pass `{ sum: false }` as a second parameter of `addTokens()` method.
```
Comment (cafebedouin) [inline docs/transaction-building.md:129]:
```suggestion
If you want it to be added at the end of the list, you can pass `{ sum: false }` as a second parameter of the `addTokens()` method.
```

File: docs/transaction-building.md [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+:::
+
+### Step. 4.2: Mint a token
+
+Tokens on Ergo can be minted, that is, created, by a transaction. To create a new token, set the `Token ID` equal to the first input's `Box ID` and add it to the outputs. Additionally, [EIP-4](https://github.com/ergoplatform/eips/blob/master/eip-0004.md) defines a pattern for uniform token minting across the ecosystem by settings values on additional registers.
+
+The `OutputBuilder`'s `mintToken()` method provides a seamless way to mint EIP-4 tokens.
+
+In the following example, we are minting a token named `TestToken` and sending it to `9gNvAv97W71Wm33GoXgSQBFJxinFubKvE6wh2dEhFTSgYEe783j` address.
```
Comment (cafebedouin) [inline docs/transaction-building.md:148]:
```suggestion
In the following example, a token named `TestToken` will be minted and and sent to the `9gNvAv97W71Wm33GoXgSQBFJxinFubKvE6wh2dEhFTSgYEe783j` address.
```

File: docs/transaction-building.md [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+Only one token can be minted per transaction.
+:::
+
+## Step. 5: Set the change address
+
+Often, all the funds included in the input boxes are not used by the transaction. An address needs to be set to receive these funds. This is referred to as the change from the eUTxO transaction.
+
+To do so, you must call the `sendChangeTo()` method and pass an address as the only parameter.
+
+In the following example, we are sending the change to the `9gNvAv97W71Wm33GoXgSQBFJxinFubKvE6wh2dEhFTSgYEe783j` address.
```
Comment (cafebedouin) [inline docs/transaction-building.md:175]:
```suggestion
In the following example, the change will be sent to the `9gNvAv97W71Wm33GoXgSQBFJxinFubKvE6wh2dEhFTSgYEe783j` address.
```

File: docs/transaction-building.md [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+new TransactionBuilder(creationHeight)
+  // ...
+  .payMinFee(); // [!code focus]
+```
+
+## Step. 7: Build
+
+Now that all pieces are put together, it's time to build the transaction and obtain the **Unsigned Transaction** object. For that you can use the `build()` method.
+
+The `build()` method must be the last called method on the `TransactionBuilder` as it will do all necessary validations, input selection, change calculations and return an unsigned transaction object which can be signed by any Ergo signing tool.
```
Comment (cafebedouin) [inline docs/transaction-building.md:208]:
```suggestion
The `build()` method must be the last called method on the `TransactionBuilder` as it will do all necessary validations, input selection, change calculations and return an unsigned transaction object which can be signed by any Ergo signing tool, such as a wallet.
```

Comment (cafebedouin) [reply docs/transaction-building.md:208]: Signing tool is inclusive, but it is also vague. I wanted to added a concrete, and typical, example of signing tool. Also, good bridge to next part, which is also one option among many, such as Ergopay signing.

File: docs/transaction-building.md [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+<!-- prettier-ignore -->
+```ts
+new TransactionBuilder(creationHeight)
+   // ...
+  .build("EIP-12"); // [!code focus]
+```
+
+## Example
+
+Building and signing a transaction using the dApp Connector protocol as a source of blockchain data.
```
Comment (cafebedouin) [inline docs/transaction-building.md:228]:
```suggestion
Build and sign a transaction using the dApp Connector protocol to access blockchain data.
```

**Review by cafebedouin — COMMENTED** [review]
Reads much better to me on this round.

File: docs/transaction-building.md (hunk 12 above)
Comment (capt-nemo429) [reply docs/transaction-building.md:17]: Yeah, good point! When I said "a set of boxes" I was referring to the transaction's data structure. But your approach is better as it covers the whole concept of a transaction, not only the data structure.

File: docs/transaction-building.md (hunk 14 above)
Comment (capt-nemo429) [reply docs/transaction-building.md:91]: Great! I will try to avoid this in future texts.

File: docs/transaction-building.md (hunk 19 above)
Comment (capt-nemo429) [reply docs/transaction-building.md:228]: I guess "to interact" fit better here, will rephrase this to be a bit more precise.
